src/NHibernate/Engine/ActionQueue.cs

```diff
@@ -505,106 +507,172 @@ private class InsertActionSorter
+						if (nextBatchIdentifier.ChildEntityNames.Contains(entityName) &&
+							!batchIdentifier.ChildEntityNames.Contains(nextBatchIdentifier.EntityName))
+						{
+							_orderedBatches.RemoveAt(i);
+							_orderedBatches.Insert(j, batchIdentifier);
+						}
 					}
 				}
-			}
 
-			private int GetBatchNumber(EntityInsertAction action, string entityName)
-			{
-				int batchNumber;
-				if (_latestBatches.TryGetValue(entityName, out batchNumber))
+				// now rebuild the insertions list. There is a batch for each entry in the batches list.
+				foreach (var batchIdentifier in _orderedBatches)
 				{
-					// There is already an existing batch for this type of entity.
-					// Check to see if the latest batch is acceptable.
-					if (IsProcessedAfterAllAssociatedEntities(action, batchNumber))
-						return batchNumber;
+					var batch = _actionBatches[batchIdentifier];
+					_actionQueue.insertions.AddRange(batch);
 				}
-				
-				// add an entry for this type of entity.
-				// we can be assured that all referenced entities have already
-				// been processed,
-				// so specify that this entity is with the latest batch.
-				// doing the batch number before adding the name to the list is
-				// a faster way to get an accurate number.
-
-				batchNumber = _actionBatches.Count;
-				_latestBatches[entityName] = batchNumber;
-				return batchNumber;
 			}
 
-			private bool IsProcessedAfterAllAssociatedEntities(EntityInsertAction action, int latestBatchNumberForType)
+			/// <summary>
+			/// Add parent and child entity names so that we know how to rearrange dependencies
+			/// </summary>
+			/// <param name="action">The action being sorted</param>
+			/// <param name="batchIdentifier">The batch identifier of the entity affected by the action</param>
+			private void AddParentChildEntityNames(EntityInsertAction action, BatchIdentifier batchIdentifier)
 			{
 				var propertyValues = action.State;
-				var propertyTypes = action.Persister.ClassMetadata.PropertyTypes;
+				// NH: it seems we support custom persister which does not provide meta-data (see CustomPersister in NHibernate.DomainModel).
+				var propertyTypes = action.Persister.ClassMetadata?.PropertyTypes;
+				if (propertyTypes == null)
+				{
+					log.WarnFormat("Entity {0} persister does not provide meta-data, its batch ordering may be wrong and cause a failure.",
+						batchIdentifier.EntityName);
+					return;
+				}
 
 				for (var i = 0; i < propertyValues.Length; i++)
 				{
 					var value = propertyValues[i];
 					var type = propertyTypes[i];
-
-					if (type.IsEntityType &&
-						value != null)
+					if (type.IsEntityType && value != null)
+					{
+						var entityType = (EntityType)type;
+						// NHibernate has non constrained one-to-one among its state while it seems Hibernate does not.
+						// We have to filter them out, since they are indeed children, not parents.
+						if (!(entityType.IsOneToOne && entityType.IsNullable))
```

Comment:
See code comment, either Hibernate has still a bug with those changes, or Hibernate does not fetch the same data in state.

Without that, I get both one-to-one sides referencing each other as parent. This causes the sorting algorithm to swap them, even if the actual parent was the first. 14 tests were failing before I change that.